src/transaction/transaction.py

```diff
@@ -0,0 +1,255 @@
+        self.outputs.append({  # add new output
+            "address": hash_address,
+            "amount": amount
+        })
+
+    def add_coinbase_output(self, address, amount):
+        """
+        Adds an output to an address without having to verify inputs
+        :param address: client's public key
+        :param amount: amount to send
+        :return: none
+        """
+        hash_address = SHA256.new(address.encode('utf-8')).hexdigest()
+
+        self.output_count += 1
+        self.outputs.append({  # add new output
+            "address": hash_address,
+            "amount": amount
+        })
+
+    def unlock_inputs(self, private_key, public_key):
+        """
+        Unlock previous unspent transaction outputs for a new transaction.
+        Compose a transaction message consisting of:
+            - input transaction ID
+            - input output index
+            - unspent transaction object's public key script (hashed pubkey)
+            - this transaction's public key script (s)?
+            - this transaction's output amount (s)?
+        and sign it with this node's private key. 
+
+        Place a dict consisting of {key: public_key, sig: signature} in the
+        corresponding input at "unlock" where `public_key` is this node's
+        full public key and `signature` is the signed transaction message.
+        """
+        if self.unused_amount != 0:  # use up all input amounts (change)
+            hash_address = SHA256.new(public_key.encode('utf-8')).hexdigest()
+            self.outputs.append({
+                "address": hash_address,
+                "amount": self.unused_amount
+            })
+            self.output_count += 1
+            self.unused_amount = 0
+
+        for tnx_input in self.inputs:  # for each input
+            utxo = find_unspent_output(tnx_input['transaction_id'],
+                                       tnx_input['output_index'],
+                                       tnx_input['block_hash'])
+            if utxo:
+                transaction_message = SHA256.new((  # compose message
+                    str(tnx_input['transaction_id']) +
+                    str(tnx_input['output_index']) +
+                    str(utxo['address']) +  # hashed public key address
+                    str(self.outputs)
+                ).encode('utf-8'))
+                signer = DSS.new(private_key, 'fips-186-3')
+                signature = signer.sign(transaction_message)
+                encoded = base64.b64encode(signature).decode()
+                unlock = {  # create unlocking portion of the transaction
+                    "public_key": public_key,
+                    "signature": encoded
+                }
+                tnx_input['unlock'] = unlock  # assign to input.
+            else:
+                # TODO: raise error
+                print("Invalid input found for {}".format(tnx_input))
+
+    def verify(self):
+        # TODO: Maybe rename this function to authenticate
```

Comment:
So in block chain context, and more specifically transaction context, _verify_ means a lot of things. I can verify if an incoming transaction is authentic, I can verify whether or not someone has enough funds, I can verify double-spent transactions. Because of this ambiguity, I think it's important to name these functions by exactly what they are doing. It becomes confusing when outside classes need to use these functions because it's hard to determine what it's verifying. It might seem useless now, but only because transaction class is not fully completed, once it is, itll make more sense. btw Authenticate may not be a good name, but it needs to be something differrnt from _verify_. Maybe something like verify_auth